util/ribbon_impl.h

```diff
@@ -0,0 +1,487 @@
+  inline bool UseBacktrack() const { return true; }
+  inline void BacktrackPut(Index i, Index to_save) { backtrack_[i] = to_save; }
+  inline Index BacktrackGet(Index i) const { return backtrack_[i]; }
+
+  // ********************************************************************
+  // Some useful API, still somewhat low level. Here an input is
+  // a Key for filters, or std::pair<Key, ResultRow> for general PHSF.
+
+  // Adds a range of inputs to the banding, returning true if successful.
+  // False means none or some may have been successfully added, so it's
+  // best to Reset this banding before any further use.
+  //
+  // Adding can fail even before all the "slots" are completely "full".
+  //
+  template <typename InputIterator>
+  bool AddRange(InputIterator begin, InputIterator end) {
+    return BandingAddRange(this, *this, begin, end);
+  }
+
+  // Adds a range of inputs to the banding, returning true if successful,
+  // or if unsuccessful, rolls back to state before this call and returns
+  // false. Caller guarantees that the number of inputs in this batch
+  // does not exceed `backtrack_size` provided to Reset.
+  //
+  // Adding can fail even before all the "slots" are completely "full".
+  //
+  template <typename InputIterator>
+  bool AddRangeOrRollBack(InputIterator begin, InputIterator end) {
+    return BandingAddRange(this, this, *this, begin, end);
+  }
+
+  // Adds a single input to the banding, returning true if successful.
+  // If unsuccessful, returns false and banding state is unchanged.
+  //
+  // Adding can fail even before all the "slots" are completely "full".
+  //
+  bool Add(const AddInput& input) { return AddRange(&input, &input + 1); }
+
+  // Return the number of "occupied" rows (with non-zero coefficients stored).
+  Index GetOccupiedCount() const {
+    Index count = 0;
+    const Index num_slots = num_starts_ + kCoeffBits - 1;
+    for (Index i = 0; i < num_slots; ++i) {
+      if (coeff_rows_[i] != 0) {
+        ++count;
+      }
+    }
+    return count;
+  }
+
+  // ********************************************************************
+  // High-level API
+
+  // Iteratively (a) resets the structure for `num_slots`, (b) attempts
+  // to add the range of inputs, and (c) if unsuccessful, chooses next
+  // hash seed, until either successful or unsuccessful with max_seed
+  // (minimum one seed attempted). Returns true if successful. In that
+  // case, use GetSeed() to get the successful seed.
+  template <typename InputIterator>
+  bool ResetAndFindSeedToSolve(Index num_slots, InputIterator begin,
+                               InputIterator end, Seed max_seed) {
+    StandardHasher<TypesAndSettings>::ResetSeed();
+    do {
+      Reset(num_slots);
+      bool success = AddRange(begin, end);
+      if (success) {
+        return true;
+      }
+    } while (StandardHasher<TypesAndSettings>::NextSeed(max_seed));
+    // no seed through max_seed worked
+    return false;
```

Comment:
Question: what's the expectation from the user if it's failed? Increase the `kNumSlots` and retry? If that's the case, a helper function to do that might be helpful.

Reply:
I think that's going to be application specific. Increasing kNumSlots could be an approach, but I expect us to fall back on Bloom filter if we find ourselves with extraordinarily bad luck, or something is broken. We should be able to make that reliably extremely rare without approaches like increasing kNumSlots.